### ISSUE TITLE
compute_waveguide_s_matrix has no checkpoint_segments — broadband waveguide-AD OOMs while the MSL path (compute_msl_s_matrix) does not

### ISSUE THREAD

**Issue by bk-squared** [bk-squared/rfx#131]
> Reported from a separate analysis workspace; **no rfx file has been committed.**
> This is a **missed port**, not a regression — the segmented-checkpoint machinery
> already exists in `run()` (#73) and was wired to the MSL path (#125); the
> waveguide entry point was simply never connected to it. The complete patch is
> inlined at the end; it is a faithful, default-off, +63/-1 two-file change.

## Summary

`Simulation.compute_waveguide_s_matrix` (`rfx/api/_sparams.py`) exposes no
`checkpoint_segments` knob, so a reverse-mode (`jax.grad`) sweep over its
`eps_override` channel stores the **entire** FDTD trajectory and peak backward
memory grows `O(n_steps · |carry|)`. The sibling microstrip-line entry point
`compute_msl_s_matrix` was wired to segmented gradient checkpointing
(`checkpoint_segments`, uniform path) in commit `6983385`
("feat(msl): thread checkpoint_segments/_every into compute_msl_s_matrix"),
and that knob was then used in **PR #125** (`0a3dd62`) to cap the MSL AD tape so
it fits a 48 GB A6000 (`checkpoint_segments=14` over `n_steps=3934`). The
waveguide path was explicitly deferred at the time — `6983385`'s own commit
message notes:

> "Only the MSL path routes through forward(); compute_waveguide_s_matrix runs a
> bespoke scan (no forward() call) so symmetric WG checkpointing is a separate,
> larger change — not done here."

In practice the "larger change" is small: the segmented-checkpoint machinery
itself lives in `rfx.simulation.run()` (added under **issue #73**, commit
`b6c4153`), and the waveguide extractors already call **that same `run()`** via
`from rfx.simulation import run as run_simulation`. They simply never forwarded
`checkpoint`/`checkpoint_segments`. The fix is to thread the existing kwargs
through; no new checkpointing logic is introduced.

## Affected API

`Simulation.compute_waveguide_s_matrix(...)` — the public reverse-mode entry
point for multiport waveguide S-parameter inverse design (`Simulation` inherits
`_SparamMixin`). `optimize()`/`topology()` do not call it directly; users wrap
it in their own `jax.grad`/`value_and_grad` objective, exactly as the MSL AD
tests do.

## Repro (OOM)

The motivating case is the rfx-tap **idx4 broadband dielectric-taper** match
study, differentiating `compute_waveguide_s_matrix(normalize=False)` w.r.t. an
`eps_override` taper profile:

- An 80 mm taper at **220 periods (~12.7k steps)** OOM-killed a **48 GB A6000**
  on the un-checkpointed waveguide AD tape.
- It was only made to run by truncating to **120 periods**, which yielded a
  +8.95 dB broadband-match margin over the best single-section reference.
- This mirrors the MSL OOM that PR #125 fixed: VESSL run `369367241162` showed
  `test_msl_s_matrix_ad_end_to_end` OOMs even a 48 GB A6000 on the
  un-checkpointed `~3934`-step tape, resolved with `checkpoint_segments=14`.

The waveguide path simply lacks the equivalent knob, so the same workaround is
unavailable.

## Proposed fix (mirrors #73 machinery + the MSL precedent)

Add `checkpoint_segments: int | None = None` to
`compute_waveguide_s_matrix` and thread it, with `checkpoint=(checkpoint_segments
is not None)`, into the three uniform extractors reachable from it:

- `extract_waveguide_s_matrix` (normalize=False),
- `extract_waveguide_s_matrix_flux` (normalize='flux'),
- `extract_waveguide_s_params_normalized` (normalize=True),

each of which calls `rfx.simulation.run(...)`. `run()` already:

- requires `checkpoint=True` to activate the segment-level
  `jax.checkpoint(segment_body, prevent_cse=False)` (passing
  `checkpoint_segments` alone is inert), and
- validates that `K` divides `n_steps`, raising `ValueError` with a suggested
  nearest divisor (padding is rejected because it would shift the V/I DFT
  accumulator windows).

No checkpoint logic is duplicated — the divisor guard and the remat are reused
verbatim from `run()`, exactly as the MSL path reuses `forward()`.

**Default-off / no behaviour change.** `checkpoint_segments=None` threads
`checkpoint=False`, so `run()` takes the legacy `jax.lax.scan(step_fn, ...)`
branch with no `jax.checkpoint` wrap.

**NU gate (mirrors the MSL/#73 uniform-only restriction).** The non-uniform
two-run extractor goes through `run_nonuniform_path`, which does not accept
`checkpoint_segments`. `compute_waveguide_s_matrix` therefore raises
`NotImplementedError` if `checkpoint_segments` is set on a mesh with
`dx_profile`/`dy_profile`, so the user is never silently dropped back to the
linear-memory scan the kwarg was meant to fix. This matches the existing
`forward(checkpoint_segments=...)` gate in `rfx/api/_execute.py`. (No
`checkpoint_every` NU counterpart is added — that was a pre-existing NU-runner
knob, not something the MSL uniform lane exposes.)

## Validation (CPU falsification — heavy GPU run owned by VESSL)

Following the MSL precedent (`6983385` falsified on a tiny thru-line on CPU and
deferred the converged run to the VESSL gate), the patch was verified on a
throwaway clone at the patch's base commit (`e2dd3b1`, current `origin/main`
HEAD; both `6983385` and #125 are in its history). Tiny WR-90 2-port, grid
`57×15×8`, `num_periods=4` → `n_steps=70`, `checkpoint_segments=7` (the divisor
of 70 nearest `√70≈8.4`), with an `eps=2.0` dielectric slab in the
`eps_override` channel, CPU, `JAX_ENABLE_X64=0` (the waveguide scan carry is
complex64):

| Check | Result |
|-------|--------|
| `git apply --check` to `e2dd3b1` | clean (exit 0) |
| forward `S(checkpoint_segments=None)` vs `S(=7)`, normalize=False/True/'flux' | `max\|ΔS\| = 0.000e+00` (bit-identical, all three modes) |
| patched(`None`) vs **pristine unpatched** `e2dd3b1`, all three modes | `max\|ΔS\| = 0.000e+00` (default unchanged) |
| `jax.grad(mean\|S\|)` w.r.t. full `eps_override`, None vs K=7 | finite, 5517/6840 cells nonzero, absmax `5.19e-4`, `max\|Δgrad\| = 3.49e-10`, rel `6.7e-7` |
| non-divisor `checkpoint_segments=8` over `n_steps=70` | raises `ValueError` (suggests K=7) |
| NU mesh (`dx_profile`) + `checkpoint_segments=7` | raises `NotImplementedError` before any run |
| existing `tests/test_waveguide_sparam_ad.py` | 10 passed (no regression) |

The converged ~150 mm idx4 broadband taper on GPU/VESSL is the production target
and is **not** part of this patch's verification — only the
checkpoint-equivalence claim is CPU-falsified here.

## Recommended follow-ups (not in this patch)

1. Add an upstream regression test mirroring
   `tests/test_scan_segmented_checkpoint.py` (forward + gradient equivalence,
   divisor enforcement, NU `NotImplementedError`) so CI guards the equivalence.
2. NU waveguide checkpointing (thread a chunk-size knob through
   `run_nonuniform_path`) — tracked as the same NU follow-up referenced by #73.

## Scope / honesty notes

- This patch touches only the **uniform** extractors reachable from
  `compute_waveguide_s_matrix`. The multimode extractors
  (`extract_multimode_*`) are left untouched, consistent with the NU gate.
- A separate, **pre-existing** tape break in the `add_waveguide_port` *builder*
  (gradient w.r.t. a scalar that rebuilds the port) is documented by
  `test_wg_smatrix_upstream_builder_tape_break_documented` and is **unaffected**
  by this patch — the `eps_override` channel used by the idx4 taper does carry
  gradient.

## Patch (`wg_smatrix_checkpoint.patch`, applies to `e2dd3b1`)

```diff
diff --git a/rfx/api/_sparams.py b/rfx/api/_sparams.py
index 3c44dd9..86dd842 100644
--- a/rfx/api/_sparams.py
+++ b/rfx/api/_sparams.py
@@ -51,6 +51,7 @@ class _SparamMixin:
         subpixel_smoothing: bool | str = False,
         eps_override: "jnp.ndarray | None" = None,
         sigma_override: "jnp.ndarray | None" = None,
+        checkpoint_segments: int | None = None,
     ) -> WaveguideSMatrixResult:
         """Compute a theoretically clean axis-normal boundary-aperture waveguide S-matrix.
 
@@ -91,6 +92,31 @@ class _SparamMixin:
             in S11 and the round-trip dispersion error in the
             ``normalize=True`` diagonal formula.  Costs 2 × N_ports
             FDTD runs (same as ``normalize=True``).
+        checkpoint_segments : int or None
+            Gradient-checkpointing segment count for the reverse-mode AD
+            tape on the **uniform** waveguide path. Mirrors the MSL knob
+            (issue #73 / PR #125): the n_steps scan is split into K
+            segments and each segment is rematerialised as a unit via
+            ``jax.checkpoint``, so backward peak memory drops from
+            ``O(n_steps·|carry|)`` to ``O((K+n_steps/K)·|carry|)`` — the
+            standard ~√n_steps trade (≈2× backward compute). ``K`` is
+            forwarded to :func:`rfx.simulation.run` (which also receives
+            ``checkpoint=True`` so the segment-level remat is active) on
+            every per-driven-port run of the selected extractor.
+
+            ``K`` must DIVIDE the (auto-computed) ``n_steps`` exactly;
+            padding is rejected by the runner because it would shift the
+            waveguide V/I DFT accumulator windows and change the result.
+            Choose the divisor nearest ``√n_steps``. Default ``None``
+            leaves the scan unchanged and is byte-identical to the
+            pre-checkpoint path (no ``jax.checkpoint`` wrap at all).
+
+            Only the uniform single-device path is wired. A non-uniform
+            mesh (``dx_profile`` / ``dy_profile``) raises
+            ``NotImplementedError`` — the NU two-run extractor runs
+            through ``run_nonuniform_path`` and does not yet thread the
+            segment knob (NU follow-up, mirrors the MSL uniform-only gate
+            in :meth:`forward`).
         """
         if not normalize:
             import warnings
@@ -133,6 +159,21 @@ class _SparamMixin:
         # is met (``normalize=True``, single-mode ports); otherwise
         # raise so the user is not given silently-wrong numbers.
         if self._dx_profile is not None or self._dy_profile is not None:
+            # Segmented checkpointing (issue #73 / PR #125) is wired only
+            # on the uniform single-device path, exactly like the MSL
+            # uniform-only gate in Simulation.forward(). The NU two-run
+            # extractor runs through run_nonuniform_path, which does not
+            # yet accept checkpoint_segments. Reject loudly so the user
+            # does not silently fall back to the linear-memory scan the
+            # kwarg was meant to fix.
+            if checkpoint_segments is not None:
+                raise NotImplementedError(
+                    "compute_waveguide_s_matrix(checkpoint_segments=...) is "
+                    "currently wired only on the uniform single-device path "
+                    "(issue #73 / PR #125). Drop checkpoint_segments or run "
+                    "on a uniform mesh (no dx_profile / dy_profile). NU "
+                    "support is tracked as a follow-up."
+                )
             unsupported = []
             if normalize is not True and normalize != "flux":
                 unsupported.append("normalize=True or normalize='flux' is required")
@@ -521,6 +562,7 @@ class _SparamMixin:
                 conformal_weights=conformal_weights,
                 aniso_inv_eps=aniso_inv_eps,
                 ref_aniso_inv_eps=ref_aniso_inv_eps,
+                checkpoint_segments=checkpoint_segments,
             )
         elif normalize:
             from rfx.core.yee import init_materials as _init_vacuum_materials
@@ -544,6 +586,7 @@ class _SparamMixin:
                 conformal_weights=conformal_weights,
                 aniso_inv_eps=aniso_inv_eps,
                 ref_aniso_inv_eps=ref_aniso_inv_eps,
+                checkpoint_segments=checkpoint_segments,
             )
         else:
             s_params = extract_waveguide_s_matrix(
@@ -560,6 +603,7 @@ class _SparamMixin:
                 aniso_eps=aniso_eps,
                 conformal_weights=conformal_weights,
                 aniso_inv_eps=aniso_inv_eps,
+                checkpoint_segments=checkpoint_segments,
             )
         reference_planes = np.array(
             [
diff --git a/rfx/sources/waveguide_port.py b/rfx/sources/waveguide_port.py
index 5cda698..5dbfddb 100644
--- a/rfx/sources/waveguide_port.py
+++ b/rfx/sources/waveguide_port.py
@@ -2037,12 +2037,22 @@ def extract_waveguide_s_matrix(
     aniso_eps: tuple | None = None,
     conformal_weights: tuple | None = None,
     aniso_inv_eps: tuple | None = None,
+    checkpoint_segments: int | None = None,
 ) -> jnp.ndarray:
-    """Assemble an x-directed waveguide S-matrix via one-driven-port-at-a-time runs."""
+    """Assemble an x-directed waveguide S-matrix via one-driven-port-at-a-time runs.
+
+    ``checkpoint_segments`` (issue #73 / PR #125): when not ``None``, each
+    per-driven-port FDTD run is launched with segmented gradient
+    checkpointing (``checkpoint=True`` + ``checkpoint_segments=K``) so the
+    reverse-mode AD tape fits ~√n_steps memory instead of O(n_steps). ``K``
+    must divide ``n_steps`` (the runner rejects a non-divisor). ``None``
+    (default) is byte-identical to the legacy linear-memory scan.
+    """
     if len(port_cfgs) < 2:
         raise ValueError(
             "extract_waveguide_s_matrix requires at least two waveguide ports"
         )
+    _wg_checkpoint = checkpoint_segments is not None
     if ref_shifts is None:
         ref_shifts = tuple(0.0 for _ in port_cfgs)
     if len(ref_shifts) != len(port_cfgs):
@@ -2091,6 +2101,8 @@ def extract_waveguide_s_matrix(
             aniso_eps=aniso_eps,
             conformal_weights=conformal_weights,
             aniso_inv_eps=aniso_inv_eps,
+            checkpoint=_wg_checkpoint,
+            checkpoint_segments=checkpoint_segments,
         )
         final_cfgs = result.waveguide_ports or ()
         if len(final_cfgs) != n_ports:
@@ -2142,6 +2154,7 @@ def extract_waveguide_s_matrix_flux(
     aniso_inv_eps: tuple | None = None,
     ref_aniso_inv_eps: tuple | None = None,
     ref_materials_per_port: "list | None" = None,
+    checkpoint_segments: int | None = None,
 ) -> jnp.ndarray:
     """Hybrid power-flux magnitude + modal phase waveguide S-matrix.
 
@@ -2230,6 +2243,8 @@ def extract_waveguide_s_matrix_flux(
         cpml_axes=cpml_axes,
         pec_axes=pec_axes,
         periodic=periodic,
+        checkpoint=checkpoint_segments is not None,
+        checkpoint_segments=checkpoint_segments,
     )
 
     for drive_idx in range(n_ports):
@@ -2334,6 +2349,7 @@ def extract_waveguide_s_params_normalized(
     ref_aniso_eps: tuple | None = None,
     aniso_inv_eps: tuple | None = None,
     ref_aniso_inv_eps: tuple | None = None,
+    checkpoint_segments: int | None = None,
 ) -> jnp.ndarray:
     """Two-run normalized waveguide S-matrix.
 
@@ -2427,6 +2443,8 @@ def extract_waveguide_s_params_normalized(
         cpml_axes=cpml_axes,
         pec_axes=pec_axes,
         periodic=periodic,
+        checkpoint=checkpoint_segments is not None,
+        checkpoint_segments=checkpoint_segments,
     )
     # ``aniso_eps`` is per-component smoothed permittivity for the device
     # geometry. The reference run is vacuum and has no ε interfaces, so
```
